crates/ra_hir/src/ty.rs

```diff
@@ -0,0 +1,617 @@
+            ImplTraitType(_inner) => Ty::Unknown,   // TODO
+            DynTraitType(_inner) => Ty::Unknown,    // TODO
+        })
+    }
+
+    pub fn unit() -> Self {
+        Ty::Tuple(Vec::new())
+    }
+}
+
+impl fmt::Display for Ty {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Ty::Bool => write!(f, "bool"),
+            Ty::Char => write!(f, "char"),
+            Ty::Int(t) => write!(f, "{}", t.ty_to_string()),
+            Ty::Uint(t) => write!(f, "{}", t.ty_to_string()),
+            Ty::Float(t) => write!(f, "{}", t.ty_to_string()),
+            Ty::Str => write!(f, "str"),
+            Ty::Slice(t) => write!(f, "[{}]", t),
+            Ty::Never => write!(f, "!"),
+            Ty::Tuple(ts) => {
+                write!(f, "(")?;
+                for t in ts {
+                    write!(f, "{},", t)?;
+                }
+                write!(f, ")")
+            }
+            Ty::FnPtr(sig) => {
+                write!(f, "fn(")?;
+                for t in &sig.input {
+                    write!(f, "{},", t)?;
+                }
+                write!(f, ") -> {}", sig.output)
+            }
+            Ty::Unknown => write!(f, "[unknown]"),
+        }
+    }
+}
+
+pub fn type_for_fn(db: &impl HirDatabase, f: Function) -> Cancelable<Ty> {
+    let syntax = f.syntax(db);
+    let node = syntax.borrowed();
+    // TODO we ignore type parameters for now
+    let input = node
+        .param_list()
+        .map(|pl| {
+            pl.params()
+                .map(|p| {
+                    p.type_ref()
+                        .map(|t| Ty::new(db, t))
+                        .unwrap_or(Ok(Ty::Unknown))
+                })
+                .collect()
+        })
+        .unwrap_or_else(|| Ok(Vec::new()))?;
+    let output = node
+        .ret_type()
+        .and_then(|rt| rt.type_ref())
+        .map(|t| Ty::new(db, t))
+        .unwrap_or(Ok(Ty::Unknown))?;
+    let sig = FnSig { input, output };
+    Ok(Ty::FnPtr(Arc::new(sig)))
+}
+
+// TODO this should probably be per namespace (i.e. types vs. values), since for
+// a tuple struct `struct Foo(Bar)`, Foo has function type as a value, but
+// defines the struct type Foo when used in the type namespace. rustc has a
+// separate DefId for the constructor, but with the current DefId approach, that
+// seems complicated.
+pub fn type_for_def(db: &impl HirDatabase, def_id: DefId) -> Cancelable<Ty> {
+    let def = def_id.resolve(db)?;
+    match def {
+        Def::Module(..) => {
+            log::debug!("trying to get type for module {:?}", def_id);
+            Ok(Ty::Unknown)
+        }
+        Def::Function(f) => type_for_fn(db, f),
+        Def::Item => {
+            log::debug!("trying to get type for item of unknown type {:?}", def_id);
+            Ok(Ty::Unknown)
+        }
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct InferenceResult {
+    type_for: FxHashMap<LocalSyntaxPtr, Ty>,
```

Comment:
We need to figure out (not in this PR) a better way to refer to things inside a function. The minimal change would be to use `LocalSyntaxPtrs`, with text offsets **relative** to the enclosing function. That way, we can reuse type-inference after most changes.

Reply:
Yeah, I was thinking about that as well, but I decided I'd just use `LocalSyntaxPtr` to get started :)

Though if we want to not have to recompute the type inference itself, we'd also have to have a position-independent representation of the function bodies, right? Like the `hir::Exprs` you mention maybe...

Reply:
>Though if we want to not have to recompute the type inference itself, we'd also have to have a position-independent representation of the function bodies, right? 

Hm, right, yeah. The current syntax tree is not even "position-based", it's "identity-based": if one parses the same file twice, the resulting trees won't be `eq`, and this actually causes [interesting behavior](https://github.com/rust-analyzer/rust-analyzer/blob/d77520fde3c953968beb09a3da80a0e7b17bbc04/crates/ra_hir/src/lib.rs#L148) because salsa relies on `Eq` a great deal. 

In general, I hope to make syntax trees a very temporarily and short-lived thing in rust-analyzer: because they are immutable and store comments & whitespace explicitly, they have a pretty-high space overhead. Ideally, only syntax trees for files opened in the editor should be present in memory, syntax trees for other files should be recreated on demand and torn down liberally. 

Using `LocalSyntaxPtr` is defenitelly the best solution for now!